lib/active_record/virtual_attributes/virtual_fields.rb

```diff
@@ -136,102 +143,70 @@ class Base
-          h = {}
-          records.each do |record|
-            # The virtual_field lookup can return Symbol/Nil/Other (typically a Hash)
-            #   so the case statement and the cases for Nil/Other are new
-
-            # each class can resolve virtual_{attributes,includes} differently
-            association = record.class.replace_virtual_fields(orig_association)
-            # 1 line optimization for single element array:
-            association = association.first if association.kind_of?(Array) && association.size == 1
-
-            case association
-            when Symbol, String
-              reflection = record.class._reflect_on_association(association)
-              next if polymorphic_parent && !reflection || !record.association(association).klass
-            when nil
-              next
-            else # need parent (preloaders_for_{hash,one}) to handle this Array/Hash
-              reflection = association
-            end
-            (h[reflection] ||= []) << record
           end
-          h
         end
-        # rubocop:enable Style/BlockDelimiters, Lint/AmbiguousBlockAssociation, Style/MethodCallWithArgsParentheses
       })
+
       class Branch
         prepend(Module.new {
+          # from branched.rb 7.0
           def grouped_records
             h = {}
             polymorphic_parent = !root? && parent.polymorphic?
             source_records.each do |record|
-              # each class can resolve virtual_{attributes,includes} differently
-              @association = record.class.replace_virtual_fields(association)
-
-              # 1 line optimization for single element array:
-              @association = association.first if association.kind_of?(Array) # && association.size == 1
-
-              case association
-              when Symbol, String
-                reflection = record.class._reflect_on_association(association)
-                next if polymorphic_parent && !reflection || !record.association(association).klass
-              when nil
-                next
-              else # need parent (preloaders_for_{hash,one}) to handle this Array/Hash
-                reflection = association
-              end
+              # begin virtual_attributes changes
+              association = record.class.replace_virtual_fields(self.association)
+              # end virtual_attributes changes
+              reflection = record.class._reflect_on_association(association)
+              next if (polymorphic_parent && !reflection) || !record.association(association).klass
+
               (h[reflection] ||= []) << record
             end
             h
           end
+
+          # branched.rb 7.0
+          def preloaders_for_reflection(reflection, reflection_records)
+            reflection_records.group_by do |record|
+              # begin virtual_attributes changes
+              needed_association = record.class.replace_virtual_fields(association)
```

Comment:
is this for fixing a test fix?  This looks like it could "fix" live code with rails 7.  I'm not sure what we weren't testing previously with rails7 and master on virtual attributes to not need this. Is it just a performance fix only, so it doesn't break without it?

Reply:
hmm... just tested.
We definitely had this before the preloader changes I did before.
So I think it did more in the past.

Now, it looks like this only fixes the 2 polymorphic tests.

```ruby
it "preloads through polymorphic (polymorphic => virtual_attribute)" do
  books = Book.includes(:author_or_bookmark => :total_books).load
  expect { expect(books.map { |b| b.author_or_bookmark.total_books }).to eq([3, 3, 3]) }.to_not make_database_queries
end
```

```
Association named 'total_books' was not found on Author; perhaps you misspelled it?
```

Reply:
ok, that makes sense.  `includes` not working wouldn't break cross repo tests but would break tests here that are specifically looking for query counts and the expected optimizations.  All good here.